### PR TITLE
feat: add types

### LINKED ISSUE #60: Add typescript type
As titled.

I want do it to improve my TS skillset

### PR DESCRIPTION
**Pull request by Eomm**
fixes #60

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+export interface FastifyOverviewDecoratorOptions {
+  /**
+   * To keep the structure light and clean, you can hide empty properties
+   * @default false
+   */
+   hideEmpty?: boolean;
+}
+
+interface OverviewStructureSource {
+  stackIndex: Number;
```
Comment (Uzlopak) [inline index.d.ts:32]: Use number instead of Number

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  fileName: String;
+  relativeFileName: String;
+  lineNumber: Number;
+  columnNumber: Number;
+  functionName: String;
+  typeName?: String;
+  methodName?: String;
```
Comment (Uzlopak) [inline index.d.ts:39]: Use string instead of String

File: test/types/intex.test.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { expectType } from 'tsd'
```
Comment (Uzlopak) [inline test/types/intex.test.ts:1]: You named this file intex.test.ts.
Name it index.test-d.ts

File: index.d.ts (hunk 1 above)
Comment (Uzlopak) [reply index.d.ts:32]: Applies everywhere where you use String

File: index.d.ts (hunk 2 above)
Comment (Uzlopak) [reply index.d.ts:39]: Applies everywhere where you use String

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  source?: OverviewStructureSource,
+  children?: OverviewStructure[],
+  decorators?: {
+    decorate: OverviewStructureDecorator[],
+    decorateRequest: OverviewStructureDecorator[],
+    decorateReply: OverviewStructureDecorator[]
+  },
+  hooks?: OverviewStructureHooks,
+  routes?: {
+    method: String,
```
Comment (Uzlopak) [inline index.d.ts:81]: You could use Pick method and url from RouteOptions and make a union type.
Or you use HttpMethods type for method

File: test/types/intex.test-d.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { expectType } from 'tsd'
+
```
Comment (Uzlopak) [inline test/types/intex.test-d.ts:2]: You still have the typo. Intex vs. index ;)

File: test/types/intex.test-d.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+const app = fastify()
+
+app
+  .register(fastifyOverview, {
+    addSource: true,
+    exposeRoute: true,
+    exposeRouteOptions: {
+      url: '/custom'
+    } as RouteOptions
```
Comment (Uzlopak) [inline test/types/intex.test-d.ts]: This is confusing me. If you do "as RouteOptions" then it means that even invalid values are handled as valid. So you should remove the "as RouteOptions". If e.g. RouteOptions has options you dont use, e.g. handler use Omit utility type. e.g. Omit<RouteOptions, 'handler'>

**Comment by Uzlopak** [comment]
Last nitpick:
You use commas in your type definitions. I assume that semicolons are the preferred way and i personally also prefer semicolons. 

The typescript documentation uses semicolons everywhere and also a google search says that:
https://www.google.com/search?q=typescript+type+comma+or+semicolon&oq=typescript+types+comma+o&aqs=chrome.1.69i57j0i22i30.7144j0j4&client=ms-android-samsung-gj-rev1&sourceid=chrome-mobile&ie=UTF-8

But that is something you can decide.

File: test/types/intex.test-d.ts (hunk 6 above)
Comment (Eomm) [reply test/types/intex.test-d.ts]: I checked a bit the ts docs and the `Partial` structure works best as `Omit` does not let the user to customize those parameters

**Pull request merged**